### PR TITLE
Add nf-test review guidelines

### PR DESCRIPTION
**Pull request by nvnieuwk**
@netlify /docs/contributing/modules

### CODE REVIEW

File: src/content/docs/contributing/modules.md [hunk 1]
```diff
@@ -1082,12 +1082,35 @@ A PR review is the process of examining a new modules' submission or the changes
 - Ensure that temporary unzipped files are removed to avoid mitigating benefits and worsening problems.
 - Ensure that large outputs are compressed with the correct tool (follow guidelines for gzip vs bzip2 vs other options).
 
-#### In `../tests/modules/nf-core/modulename/main.nf` and `../tests/modules/nf-core/modulename/meta.yml`:
+#### In `modules/nf-core/modulename/meta.yml`:
 
-- Check that there are tests for all outputs, including optional ones.
-- Check that the `meta.yml` file has correct documentation links and patterns of files.
+- Check that the file has correct documentation links and patterns of files.
 - Run the tool help and check that important input (usually optional) has not been missed.
```
Comment (mashehu) [inline src/content/docs/contributing/modules.md:1088]: that's quite an overkill imo... are people really doing this? I like more of an organic growth of inputs tbh

File: src/content/docs/contributing/modules.md [hunk 2]
```diff
@@ -1082,12 +1082,35 @@ A PR review is the process of examining a new modules' submission or the changes
...
-- Check that all outputs are captured by running pytest (e.g. on Gitpod).
+
+#### In `modules/nf-core/modulename/tests/main.nf.test`:
+
+- Check that there are tests for all outputs/inputs, including optional ones. Also check for all possible parameter variations that could lead to different outputs.
```
Comment (mashehu) [inline src/content/docs/contributing/modules.md:1092]:
```suggestion
- Check that there are tests for all inputs and outputs, including optional ones. Also check for all possible parameter variations that could lead to different outputs.
```

File: src/content/docs/contributing/modules.md [hunk 3]
```diff
@@ -1082,12 +1082,35 @@ A PR review is the process of examining a new modules' submission or the changes
...
+    - [Different assertion types](https://nf-co.re/docs/contributing/tutorials/nf-test_assertions) should be used when a straightforward `process.out` snapshot is not feasible.
```
Comment (mashehu) [inline src/content/docs/contributing/modules.md:1093]:
```suggestion
    - In general, the complete `process.out` should be snapshotted, i.e. `{ assert process.out.match() }{:groovy}`,
    - [Different assertion types](https://nf-co.re/docs/contributing/tutorials/nf-test_assertions) should be used when a straightforward `process.out` snapshot is not feasible.
```

File: src/content/docs/contributing/modules.md [hunk 4]
```diff
@@ -1082,12 +1082,35 @@ A PR review is the process of examining a new modules' submission or the changes
...
+- Check that a stub test exists for the module.
```
Comment (mashehu) [inline src/content/docs/contributing/modules.md:1094]:
```suggestion
- Check that a stub test exists for the module, and that it's test names ends in `- stub`.
```

File: src/content/docs/contributing/modules.md [hunk 5]
```diff
@@ -1082,12 +1082,35 @@ A PR review is the process of examining a new modules' submission or the changes
...
+- Check that tags for any dependent modules are specified to ensure changes to upstream modules will re-trigger tests for the current module.
+- Check that the `assertAll()` function is used and that there is at minimum a success assertion and versions in the snapshot.
```
Comment (mashehu) [inline src/content/docs/contributing/modules.md:1096]: we should lint for this. could somebody open an issue please?

File: src/content/docs/contributing/modules.md [hunk 6]
```diff
@@ -1082,12 +1082,35 @@ A PR review is the process of examining a new modules' submission or the changes
...
+- Check that the test names describe the test dataset and configuration used, some examples below:
+    - `test("homo_sapiens - [fastq1, fastq2] - bam")`
+    - `test("sarscov2 - [ cram, crai ] - fasta - fai")`
+    - `test("Should search for zipped protein hits against a DIAMOND db and return a tab separated output file of hits")`
+- Check that all input data is being referenced using the `modules_testdata_base_path` parameter:
+    - e.g. `file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)`
```
Comment (mashehu) [inline src/content/docs/contributing/modules.md:1102]:
```suggestion
    - e.g. `file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true){:groovy}`
```

File: src/content/docs/contributing/modules.md [hunk 7]
```diff
@@ -1082,12 +1082,35 @@ A PR review is the process of examining a new modules' submission or the changes
...
+
+#### In `modules/nf-core/modulename/tests/main.nf.test.snap`:
+
+- Check that all output files are present in the nf-test snapshot file, or at a minimum, check that the files exist.
+- Check that there are no md5sums representing empty files (except for the stub tests)
+
+#### Migrating pytest to nf-test:
+
+- Ensure all previous pytest assertions are converted to nf-test
+- Confirm that all pytest files are deleted
```
Comment (mashehu) [inline src/content/docs/contributing/modules.md:1112]:
```suggestion
- Ensure all previous pytest assertions are converted to nf-test.
- Confirm that all pytest files are deleted, including the entry in `tests/tags.yaml`.
```

File: src/content/docs/contributing/modules.md [hunk 8]
```diff
@@ -1082,12 +1082,35 @@ A PR review is the process of examining a new modules' submission or the changes
...
+- The module MUST use an `environment.yml` for conda environments. Previously, modules used a `conda=<name>=<version>` statement in the `main.nf` file, which has since been replaced with a dedicated `environment.yml` file.
```
Comment (mashehu) [inline src/content/docs/contributing/modules.md:1113]:
```suggestion
```
we lint for this

File: src/content/docs/contributing/modules.md [hunk 9]
```diff
@@ -1082,12 +1082,35 @@ A PR review is the process of examining a new modules' submission or the changes
...
+    - `test("homo_sapiens - [fastq1, fastq2] - bam")`
+    - `test("sarscov2 - [ cram, crai ] - fasta - fai")`
+    - `test("Should search for zipped protein hits against a DIAMOND db and return a tab separated output file of hits")`
+- Check that all input data is being referenced using the `modules_testdata_base_path` parameter:
+    - e.g. `file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)`
+
+#### In `modules/nf-core/modulename/tests/main.nf.test.snap`:
+
+- Check that all output files are present in the nf-test snapshot file, or at a minimum, check that the files exist.
+- Check that there are no md5sums representing empty files (except for the stub tests)
```
Comment (mashehu) [inline src/content/docs/contributing/modules.md:1107]:
```suggestion
```
linting takes already care of that.

File: src/content/docs/contributing/modules.md [hunk 10]
```diff
@@ -1082,12 +1082,35 @@ A PR review is the process of examining a new modules' submission or the changes
...
+- Check that the file has correct documentation links and patterns of files.
 - Run the tool help and check that important input (usually optional) has not been missed.
-- Check that all outputs are captured by running pytest (e.g. on Gitpod).
+
+#### In `modules/nf-core/modulename/tests/main.nf.test`:
+
+- Check that there are tests for all outputs/inputs, including optional ones. Also check for all possible parameter variations that could lead to different outputs.
+    - [Different assertion types](https://nf-co.re/docs/contributing/tutorials/nf-test_assertions) should be used when a straightforward `process.out` snapshot is not feasible.
+- Check that a stub test exists for the module.
+- Check that tags for any dependent modules are specified to ensure changes to upstream modules will re-trigger tests for the current module.
```
Comment (mashehu) [inline src/content/docs/contributing/modules.md:1095]:
```suggestion
```
I think we already lint for this...

**Review by mashehu — CHANGES_REQUESTED** [review]

File: src/content/docs/contributing/modules.md (hunk 1 above)
Comment (nvnieuwk) [reply src/content/docs/contributing/modules.md:1088]: That was already there :grin: do you want me to remove it?

**Comment by nvnieuwk** [comment]
Thanks for the thorough review @mashehu

File: src/content/docs/contributing/modules.md (hunk 1 above)
Comment (mashehu) [reply src/content/docs/contributing/modules.md:1088]: yes, please.

File: src/content/docs/contributing/modules.md (hunk 5 above)
Comment (nvnieuwk) [reply src/content/docs/contributing/modules.md:1096]: Done #2415

**Comment by nvnieuwk** [comment]
Prettier is complaining about the sub-points in the checklist, is there a better way to do this?

**Review by jfy133 — COMMENTED** [review]
It's there a reason why you are not using my specification version?

The specifications section is what I use for reviewing...

Ach This page is such a mess now..

**Comment by mashehu** [comment]
I would also be in favor of putting this on a new page. We could collect all checklists for example similar to the tutorials.

**Comment by jfy133** [comment]
> I would also be in favor of putting this on a new page. We could collect all checklists for example similar to the tutorials. 

I'm this close to rage quitting my TODO list and just do it.

That said iirc I thought @robsyme was tasked this before I went on parental leave... I dunno if there was a PR for that?

**Comment by nvnieuwk** [comment]
Oh I didn't know that section existed, my bad :grin:

**Comment by jfy133** [comment]
Ok, you know what, pause this PR. Imma gonna do the restructuring tomorrow once I arrive in the UK (if I don't find an existing PR).

Once it's in maybe can re-do this PR into the section

**Comment by nvnieuwk** [comment]
Okay, sorry for the confusion :grin:

**Comment by jfy133** [comment]
> Okay, sorry for the confusion :grin: 

Not your fault, this has bothered me for almost a year now 😅

**Comment by nvnieuwk** [comment]
I agree it could use some restructuring :p

**Review by mashehu — CHANGES_REQUESTED** [review]
please don't merge this before #2441

**Comment by jfy133** [comment]
I would suggest @nvnieuwk closes this makes a new PR actually :grimacing: 
Most of the content should be added to and rephrased to match: https://nf-co.re/docs/contributing/nfcore_component-specifications/modules (IMO), otherwise general guidance rather than requirements should go with the new condensed page Matthias has writtetn

**Comment by nvnieuwk** [comment]
Allright, PR on the way!

**Pull request closed without merging**